### PR TITLE
Calculating Moving averages 

### PR DESCRIPTION
**Pull request by jagathkumarks**
Calculate moving averages in Java

### CODE REVIEW

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        movingAverageCalculator.add(10);
+        assertEquals(10.0, movingAverageCalculator.getMovingAverage(), 0.001);
+        movingAverageCalculator.add(20);
+        assertEquals(15.0, movingAverageCalculator.getMovingAverage(), 0.001);
+        movingAverageCalculator.add(30);
+        assertEquals(20.0, movingAverageCalculator.getMovingAverage(), 0.001);
+    }
+
+    @Test
+    public void when_window_size_is_full_oldest_value_should_be_dropped() {
```
Comment (vinipx) [inline core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java:27]: let's remove the "_" in every word. Please make it:

whenWindowSizeIsFull_shouldDropTheOldestValue()

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        MovingAverageWithApacheCommonsMath ma = new MovingAverageWithApacheCommonsMath(2);
+        ma.add(10);
+        ma.add(20);
+        assertEquals(15.0, ma.getMovingAverage(), 0.001);
+        ma.add(30);
+        assertEquals(25.0, ma.getMovingAverage(), 0.001); // 10 dropped, 20 and 30 remain
```
Comment (vinipx) [inline core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java:33]: I don't think we need this comment here

**Review by vinipx — CHANGES_REQUESTED** [review]

File: core-java-modules/core-java-calculating-moving-averages/src/main/java/com/baeldung/ExponentialMovingAverage.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        if (alpha <= 0 || alpha > 1) {
+            throw new IllegalArgumentException("Alpha must be in the range (0, 1]");
+        }
+        this.alpha = alpha;
+        this.previousEMA = null;
+    }
+
+    public double calculateEMA(double newValue) {
+        if (previousEMA == null) {
+            // First data point, EMA is same as the value itself
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/main/java/com/baeldung/ExponentialMovingAverage.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/main/java/com/baeldung/ExponentialMovingAverage.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            previousEMA = newValue;
+        } else {
+            // Calculate EMA using the formula: EMA = alpha * newValue + (1 - alpha) * previousEMA
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/main/java/com/baeldung/ExponentialMovingAverage.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/main/java/com/baeldung/MovingAverageByCircularBuffer.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        }
+    }
+
+    public double getMovingAverage() {
+        if (count == 0) {
+            return Double.NaN;
+        }
+        double sum = 0;
+        for (int i = 0; i < count; i++) {
+            sum += buffer[i]; //buffer[(head + i) % buffer.length];
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/main/java/com/baeldung/MovingAverageByCircularBuffer.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/ExponentialMovingAverageUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.baeldung;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class ExponentialMovingAverageUnitTest {
+    @Test(expected = IllegalArgumentException.class)
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/ExponentialMovingAverageUnitTest.java:9]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/ExponentialMovingAverageUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    public void when_alpha_is_invalid_should_throw_exception() {
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/ExponentialMovingAverageUnitTest.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageByCircularBufferUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    assertEquals(20.0, ma.getMovingAverage(), 0.001);
+}
+
+@Test
+public void when_window_size_is_full_oldest_value_should_be_overwritten() {
+    MovingAverageByCircularBuffer ma = new MovingAverageByCircularBuffer(2);
+    ma.add(10);
+    ma.add(20);
+    assertEquals(15.0, ma.getMovingAverage(), 0.001);
+    ma.add(30); // Overwrites 10
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageByCircularBufferUnitTest.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
+package com.baeldung;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class MovingAverageWithApacheCommonsMathUnitTest {
+
+    @Test
+    public void when_initial_average_is_calculated_it_should_be_NAN() {
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        MovingAverageWithApacheCommonsMath movingAverageCalculator = new MovingAverageWithApacheCommonsMath(5);
+        assertEquals(Double.NaN, movingAverageCalculator.getMovingAverage(), 0.001);
+    }
+
+    @Test
+    public void when_values_are_added_average_should_update_correctly() {
```
[the remark this answers is not in the document]
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java (hunk 1 above)
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java]: Fixed.

File: core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java (hunk 2 above)
Comment (jagathkumarks) [reply core-java-modules/core-java-calculating-moving-averages/src/test/java/com/baeldung/MovingAverageWithApacheCommonsMathUnitTest.java]: Fixed.

**Review by vinipx — APPROVED** [review]

File: algorithms-modules/algorithms-miscellaneous-8/pom.xml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-math3</artifactId>
+            <version>${commons-math3.version}</version>
+        </dependency>
+    </dependencies>
+</project>
```
Comment (KevinGilmore) [inline algorithms-modules/algorithms-miscellaneous-8/pom.xml:27]: You only need to add what you need to the pom.xml and leave what's already there as-is. This change looks like you've rewritten the entire file. That could cause merge conflicts if anyone else is working in the same module.

File: algorithms-modules/pom.xml [hunk 12]
```diff
@@ -25,6 +25,7 @@
         <module>algorithms-searching</module>
         <module>algorithms-sorting</module>
         <module>algorithms-sorting-2</module>
+        <module>algorithms-miscellaneous-8</module>
```
Comment (KevinGilmore) [inline algorithms-modules/pom.xml:28]: Why would you move this line? There should be no reason to modify this file.

**Review by KevinGilmore — COMMENTED** [review]
Unnecessary modifications made to pom.xml files

**Pull request closed without merging**